Partial fanout

vrom911 opened the pull request:
Part of https://github.com/cardano-scaling/hydra/issues/1468

## Summary

This PR implements the full **partial fanout** mechanism for large UTxO sets, introducing two key additions:

### `FanoutProgressDatum` — slim intermediate datum

After the first `PartialFanout` step, the head output no longer carries the full `ClosedDatum` (11 fields). Instead it carries a `FanoutProgressDatum` with only the 4 fields needed for subsequent steps: `headId`, `parties`, `contestationDeadline`, `accumulatorCommitment`. This reduces datum size and script execution cost for every intermediate step.

### `FinalPartialFanout` redeemer — explicit terminal step

A dedicated `FinalPartialFanout` redeemer handles the `FanoutProgress → Final` transition. It burns all head tokens (ST + PTs) and distributes the last batch of UTxOs, making the terminal step type-safe and distinct from the intermediate `PartialFanout` redeemer.

### Updated state machine

```
Closed ──PartialFanout──────────────────────► FanoutProgress
FanoutProgress ──PartialFanout──────────────► FanoutProgress  (repeat)
FanoutProgress ──FinalPartialFanout─────────► Final           (burns tokens)
Closed ──Fanout──────────────────────────────► Final           (unchanged, ≤ threshold UTxOs)
```

### EIP-4844 KZG trusted setup

The accumulator uses the [EIP-4844 KZG trusted setup](https://ceremony.ethereum.org/) (`trusted_setup.json`) as the Common Reference String. G1 points are used on-chain (compressed BLS12-381 G1 element stored in the datum), and G2 points are used for the membership proof verification. The KZG CRS is delivered to the on-chain validator via a dedicated `vCRS` reference script output.

### `accumulatorHash` binding

`Close` and `Contest` validators bind the `accumulatorHash` (stored in `ClosedDatum`) to the G1 accumulator commitment. This ensures consistency between the signed hash and the on-chain commitment point.

### CRS validator (`vCRS`)

New `vCRS` script that holds the G2 CRS points as an inline datum, referenced by the `PartialFanout` and `FinalPartialFanout` redeemers to avoid re-embedding the large CRS in every transaction.

### cardano-api 10.21

Updated `cardano-api` dependency from `^>=10.18` to `^>=10.21`, integrating all changes from the upstream `capi-10.21` branch. This was previously carried as a `source-repository-package` override and is now resolved to the released version.

## Changes

**On-chain (`hydra-plutus`)**
- `HeadState.hs`: new `FanoutProgressDatum` type; `FanoutProgress` state variant; `FinalPartialFanout` redeemer variant
- `Head.hs`: validator cases for `(FanoutProgress, PartialFanout)` and `(FanoutProgress, FinalPartialFanout)`; first `PartialFanout` step now outputs `FanoutProgress` datum instead of `Closed`
- `HeadError.hs`: new error codes for fanout progress transitions
- `CRS.hs`: new CRS reference script validator
- Updated `vHead.plutus`, `mHead.plutus`, `vCRS.plutus` golden scripts

**Off-chain (`hydra-tx`)**
- `Fanout.hs`: `partialFanoutTx` outputs `FanoutProgress` datum; new `finalPartialFanoutTx` that reads `FanoutProgressDatum` and burns tokens; new `observeFinalPartialFanoutTx`
- `Accumulator.hs`: `buildFromSnapshotUTxOs`, `getAccumulatorHash`, `createMembershipProofFromUTxO` and related helpers
- `KZGTrustedSetup.hs`: EIP-4844 trusted setup integration

**Node logic (`hydra-node`)**
- `HeadLogic.hs`: `emitNextFanoutStep` updated to choose between `partialFanoutTx`, `intermediatePartialFanoutTx`, and `finalPartialFanoutTx` based on remaining UTxO count and current head state
- `Chain.Direct.Handlers.hs`: observation of `FinalPartialFanout` (surfaced as `OnFanoutTx`)
- `Chain.Direct.State.hs`: `FanoutProgressState` in chain state; generators for intermediate and final partial fanout

**Tests**
- New contract tests: `PartialFanout.hs`, `FinalPartialFanout.hs`
- `KZGTrustedSetupSpec.hs`: property tests for CRS point counts and trusted setup round-trips
- `StateSpec.hs`, `HeadLogicSpec.hs`, `BehaviorSpec.hs`, `ChainObserverSpec.hs`, `HandlersSpec.hs`, `TxSpec.hs`: updated for new transitions
- `SecurityScenarios.hs`: filled missing `accumulator`, `accumulatorHash`, `accumulatorCommitment` fields post-rebase

---

* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [ ] No new TODOs introduced or explained hereafter

vrom911 committed:
Initial work on SignableRepresentation instance of Snapshot

vrom911 committed:
Initial work on SignableRepresentation instance of Snapshot

vrom911 committed:
Introduce HasAccumulatorElement class

vrom911 committed:
Introduce HasAccumulatorElement class

vrom911 committed:
Changeserialisation and signature verification

Fix plutus tests

Fix tests

vrom911 committed:
Changeserialisation and signature verification

Fix plutus tests

Fix tests

vrom911 committed:
Add accumulator to snapshot

Remove utxoHash from Snapshot

Remove leftover fn from IsTx class

vrom911 committed:
Add accumulator to snapshot

Remove utxoHash from Snapshot

Remove leftover fn from IsTx class

v0d1ch commented on hydra-cluster/test/Test/BlockfrostChainSpec.hs:
```diff
@@ -103,15 +104,22 @@ spec = around (onlyWithBlockfrostProjectFile . showLogsOnFailure "BlockfrostChai
                   { headId
                   , number = 1
                   , utxo = someUTxO
+                  , utxoHash = hashUTxO someUTxO
```
@vrom911 I don't think we want to store utxoHash here. Accumulator (and potentially CRS depending if it is dynamic or not) needs to be there so the participants sign it and it is also checked on-chain.

vrom911 committed:
Minimise golden test changes

vrom911 committed:
Minimise golden test changes

vrom911 committed:
Remove crs from Snapshot

vrom911 committed:
Remove crs from Snapshot

vrom911 committed:
Simplify golden test diff

vrom911 committed:
Simplify golden test diff

vrom911 committed:
Regenerate scripts{

Fix

vrom911 committed:
Regenerate scripts{

Fix

vrom911 committed:
Add useful fns, change the way we create acc

Temporary add acc fn to weeder ignore

vrom911 committed:
Add useful fns, change the way we create acc

Temporary add acc fn to weeder ignore

vrom911 committed:
Create BLS accumulator benchmartks with UTxO

Create BLS accumulator benchmarks with UTxO

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Create BLS accumulator benchmartks with UTxO

Create BLS accumulator benchmarks with UTxO

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Use new haskell-accumulator-1.1.0.0

vrom911 committed:
Use new haskell-accumulator-1.1.0.0

v0d1ch commented on hydra-cluster/test/Test/DirectChainSpec.hs:
```diff
@@ -312,6 +313,7 @@ spec = around (showLogsOnFailure "DirectChainSpec") $ do
             aliceChain `observesInTime` OnCollectComTx{headId}
             let v = 0
             let snapshotVersion = 0
+                accumulator = Accumulator.buildFromUTxO someUTxO
```
Each time you invoke `buildFromUTxO` you probably want to include any utxo to commit/decommit since these UTxO can also be part of the fanout outputs.

v0d1ch commented on hydra-node/src/Hydra/API/ServerOutput.hs:
```diff
@@ -331,7 +331,7 @@ instance Arbitrary NetworkInfo where
 
 -- | Get latest confirmed snapshot UTxO from 'HeadState'.
-getSnapshotUtxo :: Monoid (UTxOType tx) => HeadState tx -> Maybe (UTxOType tx)
+getSnapshotUtxo :: IsTx tx => HeadState tx -> Maybe (UTxOType tx)
```
Why did we have to add `IsTx` constraint here?

vrom911 replied:
Now `getSnapshot` requires `IsTx` cause it also `build`s the accumulator

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1116,6 +1119,7 @@ onOpenClientClose st =
 --
 -- __Transition__: 'OpenState' → 'ClosedState'
 onOpenChainCloseTx ::
+  IsTx tx =>
```
Same question regarding `IsTx` constraint?

v0d1ch commented on hydra-tx/src/Hydra/Tx/Accumulator.hs:
```diff
@@ -0,0 +1,192 @@
+      -- Compress the Point2 to a ByteString and encode as hex
+      let proofBytes = blsCompress proof
+          proofHex = decodeUtf8 $ Base16.encode proofBytes
+       in "Success: 0x" <> proofHex
```
I don't think we want to have this text pre-pended to the resulting hex? I know we are not using this function currently but perhaps returning Either is more sensible? No big issue, we can also deal with this function once we actually start to use it.

vrom911 replied:
Yes, that I was basically using for debug only 🙏🏼 
Good catch!

v0d1ch commented on hydra-tx/src/Hydra/Tx/Close.hs:
```diff
@@ -143,7 +144,12 @@ closeTx scriptRegistry vk headId openVersion confirmedSnapshot startSlotNo (endS
+   where
+    snapshot = getSnapshot confirmedSnapshot
+    snapshotUtxo = Hydra.Tx.utxo snapshot
+    closedAccumulatorHash = Accumulator.getAccumulatorHash $ Accumulator.buildFromUTxO snapshotUtxo
```
Just wanted to re-iterate once more: We need to have any utxo to commit/decommit in the accumulator here. For example: It can happen that the head gets closed in the middle of decommitting some UTxO . So since this UTxO was never brought back into L1 by the decommit tx we need to make sure fanout will do that in the end.

v0d1ch commented on hydra-tx/test/Hydra/Tx/Contract/Contest/Healthy.hs:
```diff
@@ -129,6 +131,7 @@ healthyClosedState =
       , headId = toPlutusCurrencySymbol testPolicyId
       , contesters = []
       , version = toInteger healthyCloseSnapshotVersion
+      , accumulatorHash = toBuiltin ("" :: ByteString) -- TODO: Proper accumulator hash
```
Leftover todo here

v0d1ch commented on hydra-tx/test/Hydra/Tx/Contract/FanOut.hs:
```diff
@@ -95,6 +95,7 @@ healthyFanoutDatum =
       , headId = toPlutusCurrencySymbol testPolicyId
       , contesters = []
       , version = 0
+      , accumulatorHash = toBuiltin ("" :: ByteString) -- TODO: Proper accumulator hash
```
Leftover todo here

v0d1ch reviewed (commented):
lgtm. I think the main concern is that we want to have any UTxO to commit/decommit in there when building the accumulator and I gave the reason in one of the comments.

vrom911 committed:
Add toCommit toDecommit utxo in accumulator, address review

Fix the failing observer test

Related log book entry is here https://github.com/cardano-scaling/hydra/wiki/Logbook-2025-H1#2025-12-01

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Add toCommit toDecommit utxo in accumulator, address review

Fix the failing observer test

Related log book entry is here https://github.com/cardano-scaling/hydra/wiki/Logbook-2025-H1#2025-12-01

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
WIP: Start with the changes needed for on-chain check

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
WIP: Start with the changes needed for on-chain check

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fork cardano-api until upstream is fixed

cardano-api will remove orphan instances already present in
cardano-ledger but until then we will maintain this fork so we can
progress further.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fork cardano-api until upstream is fixed

cardano-api will remove orphan instances already present in
cardano-ledger but until then we will maintain this fork so we can
progress further.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Produce correct Element from UTxO

Hash TxOuts to get a 32 bytes string length

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Produce correct Element from UTxO

Hash TxOuts to get a 32 bytes string length

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use G1 crs for on-chain check

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use G1 crs for on-chain check

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump to cardano-api 10.20 and remove src-package

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump to cardano-api 10.20 and remove src-package

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump plutus to 1.53

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump plutus to 1.53

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use patched cardano-api src package for now

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump to plutus 1.55

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump to plutus 1.55

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Setup for the on-chain checks

We want to get to a working state with all tests green before we
tackle using crs string as a reference input.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Setup for the on-chain checks

We want to get to a working state with all tests green before we
tackle using crs string as a reference input.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce CRS validator

This should be used as a reference input to reduce the execution
of contracts that need CRS string.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Hook new crs validator to script registry

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Remove CRS from Fanout redeemer

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce CRS validator

This should be used as a reference input to reduce the execution
of contracts that need CRS string.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Hook new crs validator to script registry

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Remove CRS from Fanout redeemer

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use TxOutRef to find crs datum and hook everything on-chain

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use TxOutRef to find crs datum and hook everything on-chain

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add needed datum in fanout test, keep the datum as
[BuiltinBLS12_381_G1_Element]

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add needed datum in fanout test, keep the datum as
[BuiltinBLS12_381_G1_Element]

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Keep defaultItems at 10, remove the membership check for now

I need to investigate why execution units are blowing up.
It could be a bug also.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Keep defaultItems at 10, remove the membership check for now

I need to investigate why execution units are blowing up.
It could be a bug also.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add genesis-dijkstra.json to git

Remove sha-256 since it causes nix-flake check to fail

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Fix golden files

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Fix script registry test

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Fix accumulator bench

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Add genesis-dijkstra.json to git

Remove sha-256 since it causes nix-flake check to fail

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Fix golden files

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Fix script registry test

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

Fix accumulator bench

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Optimize CRS contract

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Optimize CRS contract

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Correctly produce large enough CRS

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Correctly produce large enough CRS

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce CRS script into tx-cost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

nix flake check

Introduce CRS script into tx-cost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Introduce CRS script into tx-cost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

nix flake check

Introduce CRS script into tx-cost

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Master rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Master rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Make model spec pass by reducing the number of fanout outputs

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Make model spec pass by reducing the number of fanout outputs

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Use the updated plutus-accumulator

We need to not use bls multiplication rn

vrom911 committed:
Align accumulator element hashing and CRS use

Apply blake2b_224 before scalar conversion
Compute commitments with the actual CRS used for proofs
Build the accumulator using fanout output order

vrom911 committed:
Align accumulator element hashing and CRS use

Apply blake2b_224 before scalar conversion
Compute commitments with the actual CRS used for proofs
Build the accumulator using fanout output order

vrom911 committed:
Update plutus scripts

vrom911 committed:
Fix bench tx-cost

vrom911 committed:
Fix bench tx-cost

vrom911 committed:
Resolve warnings

vrom911 committed:
Resolve warnings

vrom911 committed:
Add adjusted redeemers to mkScriptIntegrity

vrom911 committed:
Increase memory and steps exec limits

vrom911 committed:
Increase memory and steps exec limits

vrom911 committed:
Improve checkFanout function

Optimizations applied:
 * Single-pass serialization: each 'TxOut' is serialized once via
   @serialiseData . toBuiltinData@ and reused for both UTxO hash checks
   and accumulator scalar computation.
 * Combined CRS lookup: the reference input for the CRS is looked up once
   and decoded only when the membership check is actually reached.

vrom911 committed:
Improve checkFanout function

Optimizations applied:
 * Single-pass serialization: each 'TxOut' is serialized once via
   @serialiseData . toBuiltinData@ and reused for both UTxO hash checks
   and accumulator scalar computation.
 * Combined CRS lookup: the reference input for the CRS is looked up once
   and decoded only when the membership check is actually reached.

vrom911 committed:
Fix memory budget regression in checkFanout

vrom911 committed:
Fix memory budget regression in checkFanout

vrom911 committed:
Fix warnings, rebase

vrom911 committed:
Fix warnings, rebase

v0d1ch committed:
Bring back correct execution units and make the tests pass.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bring back correct execution units and make the tests pass.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Rebase

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Unify the crs size calc in Close and Contest

vrom911 committed:
Unify the crs size calc in Close and Contest

vrom911 committed:
Fix publishHydraScriptsAs ADA amount necessary

vrom911 committed:
Fix publishHydraScriptsAs ADA amount necessary

vrom911 committed:
Fix CRS size mismatch and fanout E2E tests

vrom911 committed:
Fix CRS size mismatch and fanout E2E tests

vrom911 committed:
Try to fix tests

vrom911 committed:
Try to fix tests

vrom911 committed:
Fix accumulator benchmark to measure actual computation

vrom911 committed:
Fix accumulator benchmark to measure actual computation

vrom911 commented:
With the latest commit, I fixed the benchmarks of the accumulator.

The previous bench had a lazy evaluation bug (measuring already-evaluated thunks). Here are the **real** numbers:

### Build Accumulator from UTxO

| UTxOs  | Time   |
|--------|--------|
| 10     | 28 μs  |
| 100    | 284 μs |
| 1,000  | 3.3 ms |
| 5,000  | 20 ms  |
| 10,000 | 45 ms  |

### Create Membership Proofs (off-chain)

| Subset → Full      | Time     |
|---------------------|----------|
| 5 from 50           | 1.16 ms  |
| 10 from 100         | 1.73 ms  |
| 50 from 500         | 5.76 ms  |
| 100 from 1,000      | 10.2 ms  |
| 500 from 5,000      | 47.6 ms  |
| 1,000 from 10,000   | 91.7 ms  |

### CRS Generation (one-time startup cost)

| Size   | Time   |
|--------|--------|
| 100    | 991 ns |
| 1,000  | 10 μs  |
| 10,000 | 104 μs |

### End-to-End Simulation

| Operation                         | Time   |
|-----------------------------------|--------|
| Full snapshot cycle (100 UTxOs)   | 321 μs |
| Full snapshot cycle (1,000 UTxOs) | 4.8 ms |
| Partial fanout: 100 from 1,000   | 154 ms |

## Meaning

- **Build accumulator** scales linearly. 3.3ms for 1,000 UTxOs is good.
- **Membership proofs** are the most expensive off-chain op (~10ms for 100-from-1000), growing roughly linearly. The 154ms partial fanout end-to-end includes CRS generation + accumulator build + proof creation.
- **CRS generation** is negligible (μs range).
- **On-chain fanout** is currently capped at **~19 UTxOs** (memory-limited at 97% of 14M units). Main cost: BLS `checkMembership` (2 Miller loops + O(n²) polynomial expansion + O(n) scalar muls) plus three redundant hash checks alongside the accumulator. Removing redundant hashes + enabling the commented-out `multiScalarMul` builtin (as discussed with Thomas) could push to ~25-30 UTxOs.

vrom911 committed:
Partial fanout implementation

vrom911 committed:
Partial fanout implementation

vrom911 committed:
Fix tests

Fix warnings

vrom911 committed:
Fix tests

Fix warnings

vrom911 committed:
Add benchmarks and fix has of accumulator

vrom911 committed:
Add benchmarks and fix has of accumulator

vrom911 committed:
Remove accumulatorHash and proof from ClosedDatum

vrom911 committed:
Remove accumulatorHash and proof from ClosedDatum

vrom911 committed:
Use cardano-api 10.21 proper version

Got all changes from the capi-10.21 branch here

vrom911 committed:
Use cardano-api 10.21 proper version

Got all changes from the capi-10.21 branch here

vrom911 committed:
Update Chap, use plutus-accumulator

vrom911 committed:
Update scripts, fix tests

vrom911 committed:
Update scripts, fix tests

vrom911 committed:
remove unnecessary constraints

vrom911 committed:
Use EIP-4844 KZG trusted setup with G1 accumulator commitments

vrom911 committed:
Add value conservation check to partial fanout validator

checkPartialFanout now verifies that headInValue equals headOutValue plus the sum of all distributed output values, preventing Ada theft via outputs not counted by the membership proof. The off-chain partialFanoutTx builder is updated accordingly to subtract distributed values from the continuing head output.

vrom911 committed:
Remove unused PartialFanoutWrongProof

vrom911 committed:
Accumulate partial fanout UTxOs with correct L1 TxIns in ClosedState

vrom911 committed:
Fix tx-cost benchmarks

vrom911 committed:
Refactor shared fanout steps

vrom911 committed:
Fix api.yaml

vrom911 committed:
Add partial fanout boundary condition and mutation tests

vrom911 committed:
Use json file setup

vrom911 committed:
CHange the fanout chunk size limit

vrom911 committed:
Switch back to G1 points

vrom911 committed:
Refactor and add some more tests

vrom911 committed:
Unify impl of buildFromSnapshotUTxOs with others

vrom911 committed:
Bind accumulatorHash to G2 commitment in Close and Contest validator

vrom911 committed:
Fix era warning

vrom911 committed:
Swap G1/G2 on/off chain

vrom911 committed:
Split FanoutProgressDatum and FinalPartialFanout redeemer

    Closed ──PartialFanout──────────► FanoutProgress
    FanoutProgress ──PartialFanout──► FanoutProgress  (repeat)
    FanoutProgress ──FinalPartialFanout──────────────► Final (burns tokens)
    Closed ──Fanout──────────────────────────────────► Final (≤ threshold, unchanged)

  Off-chain, HeadLogic.emitNextFanoutStep now distinguishes FreshFanout from
  FanoutInProgress to select the correct tx builder at each step.

vrom911 committed:
Fix after rebase

v0d1ch committed:
Failing mutation proving attacker could use fake crs reference input

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Patch Head validator by parametarizing it with exspected CRS script hash

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Expose duplicate UTxO accumulator deduplication bug

  Add healthyPartialFanoutTxWithDuplicates — a partial fanout tx where two
  distinct TxIns carry identical TxOut content. The off-chain accumulator
  silently deduplicates them to one element while on-chain scalar computation
  does not, causing a pairing check mismatch. The added property currently
  fails, proving the bug exists.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix accumulator commitment to account for element multiplicity

  getAccumulatorCommitment was ignoring the Count field in the
  accumulator map.
  on-chain pairing check uses the actual duplicate scalars.

  Fix by expanding each (hash, count) entry into `count` copies
  before building the polynomial. Also fix requiredCRSPointCount
  and accumulatorSize to use total element count (sum of counts)
  rather than unique key count, so CRS sizing and the
  maxAccumulatorSize guard stay consistent.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Regenerate scripts

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Fix CPU budget regression in fanout and add value conservation to final partial fanout

  Remove the redundant KZG membership proof from the regular Fanout path since
  the UTxO hash checks already fully pin all distributed outputs. The CRS
  deserialization and BLS pairings were causing the CPU budget to be exceeded.

  Add mustConserveValue to checkFinalPartialFanout to ensure all head input
  value is accounted for by distributed outputs and burned tokens in the final
  step.

  Extract resolveCRS helper to deduplicate CRS lookup logic.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Regenerate scripts

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Regenerate scripts

v0d1ch committed:
Fix check ordering in final partial fanout and add value conservation

  Reorder checkFinalPartialFanout so the KZG membership check runs before
  mustConserveValue: changing an output value breaks membership (H60) while
  omitting outputs passes membership but fails value conservation (H4).
  Each mutation now triggers its intended error.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Regenerate scripts

v0d1ch committed:
Remove unused accumulator from fanout tx

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Bump to plutus 1.63

vrom911 committed:
Regenerate scripts

vrom911 committed:
Use geq instead of == in FinalPartialFanout value conservation check

v0d1ch committed:
Guard against zero-output final partial fanout fund theft

  With numberOfPartialOutputs = 0 the KZG check degenerates to
  e(A,G2) = e(proof,G2), which any observer can satisfy by supplying
  proof = accumulatorCommitment (public from the datum). Combined with
  the geq value check, this allowed a third party to call FinalPartialFanout
  after the contestation deadline, burn the head tokens, and route all
  remaining ADA to themselves with no outputs distributed.

  Add mustHaveOutputs (H62) as the first check in checkFinalPartialFanout.
  Update mutation tests: add MutateFinalPartialFanoutZeroOutputs using the
  exact attack proof, fix MutateFinalPartialFanoutOutputCount to use a
  non-zero wrong count so it still exercises the membership path.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Regenerate scripts

vrom911 committed:
Add completeness check to final partial fanout membership proof

The final partial fanout value check used strict equality (==) between head input value and the sum of distributed outputs plus burned tokens. On real networks the head UTxO
    carries extra ADA (min-UTxO overhead from initialization) that is not part of the L2 UTxO set, causing the check to fail with HeadValueIsNotPreserved.

    Relaxing to geq fixes the E2E failure but opens a steal-ADA vector: an attacker can supply a valid KZG subset proof for n-1 of n UTxOs, satisfying both the pairing check and the
    loose value check while leaving the nth UTxO unaccounted.

    The fix adds a completeness requirement to checkCRSAndMembership: when all remaining accumulator elements are distributed the quotient polynomial Q(X) = A(X)/A(X) = 1, so its
    commitment Q(τ)·G1 must equal the G1 generator. Any proof for a proper subset produces a non-trivial quotient, failing this check with FinalPartialFanoutMembershipFailed. The
    MutateFinalPartialFanoutStealAda mutation test is updated accordingly.

vrom911 committed:
remove snapshotAccumulator from api.json

vrom911 committed:
Add test for partial fanout resumption after insufficient funds

vrom911 committed:
Benchmark partial fanout with native-token outputs

vrom911 committed:
Remove TODO with error

vrom911 committed:
Silence stale partial fanout errors in multi-node race

ch1bo reviewed (changes requested):
Started a thorough review on this work, but only covered crypto and on-chain parts so far. Off-chain not reviewed yet.

Lots of great progress, but I'm afraid that the CRS via datum + the const true validator is a show-stopper (see "Must" comments). There is quite some potential for simplification too, so I hope this review helps already.

v0d1ch reviewed (commented):
Looking good @vrom911 I think we have one important thing to look at regarding the `crsValidator` so we should sync on this - just wanted to go over the code and give some suggestions.

v0d1ch committed:
Add explanatory comment to crsValidator returning False

  The CRS UTxO is published at an AlwaysFails script address, making it
  unspendable by the ledger rules regardless of the validator body.
  Returning False is therefore not strictly necessary but documents the
  intent: this is a permanently locked datum carrier, not an incomplete
  validator.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Replace error calls in KZGTrustedSetup with Either

  Introduces KZGSetupError — an ADT with constructors for each failure
  mode (IntegrityCheckFailed, JsonDecodeFailed, JsonParseFailed,
  InvalidG1Point, InvalidG2Point) — and threads it through all fallible
  functions in the module via Either do-notation.

  Use Set.splitAt and remove outdated comment.

v0d1ch committed:
Revert Scenarios.hs diff - not needed

Fix a test expectation

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

v0d1ch committed:
Regenerate scripts

ch1bo reviewed (approved):
I like that we have a much higher and definite upper bound on the utxo set size now. With this, we cannot "get stuck" indefinitely and could at least bring back all outputs _small enough_ to fit an individual partial fanout tx (by hand if needed).

This equates to a MASSIVE improvement of safety and liveness of the Head protocol.

NOTE: There are still absolutely too big outputs, which would be good to know that the ceiling is, but these can't make unrelated outputs stay stuck.

IMPORTANT: there is _at least_ the one fragility in the head logic that can make a node lose track of UTxOs to fanout. This can be used as an attack vector to impact safety of funds _by anyone_ because fanout validators are not authorized to only participants (that's actually a feature, but problematic here).

noonio committed:
Review feedback

noonio committed:
Todo about add to github comment benchmark thing

noonio committed:
Note about unit test

noonio committed:
hash hash hash hash hash

noonio committed:
Note about accumulatorHash

vrom911 committed:
Distinguish FinalPartialFanout from Fanout in HeadObservation

vrom911 committed:
Sweep outputs-per-step in partial fanout benchmarks instead of total UTxO count

vrom911 committed:
Move accumulator construction inside finalPartialFanoutTx

vrom911 committed:
Fix warnings

v0d1ch committed:
Track partial fanout remaining outputs by content, not by count

  The previous implementation used a count-based positional split to
  compute remaining UTxOs after a partial fanout.

  This is fragile: any participant can construct a valid on-chain partial
  fanout that distributes a non-prefix subset of outputs. The on-chain
  accumulator validates the batch without enforcing prefix order, but the
  off-chain code would then compute the wrong remaining set.

  Fix by tracking outputs as Set (TxOutType tx) throughout the partial
  fanout path and computing remaining via content-based set difference.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Refactor emitNextFanoutStep to use where-bound postChainTx with guards

vrom911 committed:
Remove test that is not needed anymore

vrom911 committed:
Fix mixed Scenario tests

vrom911 committed:
Remove toPairList function

v0d1ch committed:
Move accumulatorHash from redeemers into OpenDatum

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Fix mutation tests

v0d1ch committed:
Fix close tx generator: sync OpenDatum.accumulatorHash with snapshot

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>

vrom911 committed:
Fix checkClose to derive accumulator hash from closed datum

OpenDatum.accumulatorHash is only updated by on-chain Increment and
  Decrement transactions. Regular L2 snapshots are off-chain and never
  touch the Open datum, so the field is stale for any head that has
  processed L2 transactions.

vrom911 committed:
Warm up KZG trusted setup before opening API server socket

noonio committed:
Fix weeder

vrom911 committed:
Remove unused imports

vrom911 committed:
Fix Scenario tests

v0d1ch committed:
Restore KZG warm-up before opening API server socket

  Commit adb260894 removed the warm-up while fixing an unrelated scenario
  test ordering bug. Without it, the first call to `initTx` lazily triggers
  evaluation of all 4096 BLS12-381 G1 point decompressions inside the
  hydra-node subprocess. With blockTime=0.1s the bench-e2e only allows
  5×0.1=0.5s for HeadIsOpen; under CPU contention from three concurrent
  subprocesses the KZG evaluation exceeds that window and the waitMatch
  times out.

  Forcing evaluation before the API socket opens ensures the cryptographic
  setup is complete before any client can connect.

Signed-off-by: Sasha Bogicevic <sasha.bogicevic@iohk.io>
